### PR TITLE
[DM-33438] Clean up run_demo.sh

### PR DESCRIPTION
**Pull request by cbanek**

### CODE REVIEW

**Comment by cbanek** [comment]
This script works if I take out the multiprocessing stuff, but with the multiprocessing, I get the following error from the python tests:

___________________________________________________________________ PiplinesCheckTestCase.testExecutionButler (chain='demo_collection_exe') ___________________________________________________________________

self = <test_butler.PiplinesCheckTestCase testMethod=testExecutionButler>

    def testExecutionButler(self):
        """Check outputs match in both runs."""
    
        for chain in (EXE_CHAIN, QBB_CHAIN):
            with self.subTest(chain=chain):
                # Check that we have identical datasets in both collections
                # except for the dataset.id
                main_datasets = self._get_datasets_from_chain(MAIN_CHAIN)
                datasets = self._get_datasets_from_chain(chain)
                self.assertGreater(len(datasets), 0)
>               self.assertEqual(len(main_datasets), len(datasets))
E               AssertionError: 21 != 13

tests/test_butler.py:101: AssertionError
___________________________________________________________________ PiplinesCheckTestCase.testExecutionButler (chain='demo_collection_qbb') ___________________________________________________________________

self = <test_butler.PiplinesCheckTestCase testMethod=testExecutionButler>

    def testExecutionButler(self):
        """Check outputs match in both runs."""
    
        for chain in (EXE_CHAIN, QBB_CHAIN):
            with self.subTest(chain=chain):
                # Check that we have identical datasets in both collections
                # except for the dataset.id
                main_datasets = self._get_datasets_from_chain(MAIN_CHAIN)
                datasets = self._get_datasets_from_chain(chain)
                self.assertGreater(len(datasets), 0)
>               self.assertEqual(len(main_datasets), len(datasets))
E               AssertionError: 21 != 13

tests/test_butler.py:101: AssertionError

File: bin/run_demo.sh [hunk 1]
```diff
@@ -102,35 +102,35 @@ pipetask qgraph -b DATA_REPO/butler.yaml \
...
-pipetask --long-log run -b "$tmp_butler" --output-run "$exerun" --qgraph "$graph_file" --qgraph-node-id $node --skip-init-writes --extend-run --clobber-outputs --skip-existing
+for NODE in $(pipetask qgraph -b "$TMP_BUTLER" -g "$graph_file" --show-qgraph-header \
+    |jq -r 'first(.Nodes)[][0]')
+do
+  # Run the execution butler in multiple steps, ensuring that a fresh
+  # butler is used each time.
+  TMP_BUTLER_NODE="./tmp_execution_butler-$NODE"
+  refresh_butler $TMP_BUTLER_NODE
+
+  # Run these three pipetasks concurrently
```
Comment (timj) [inline bin/run_demo.sh:126]: These three have to run serially. The outputs of the first become inputs to the second. This set of 3 running serially can be run concurrently with the other set of 3 running serially below.

**Comment by timj** [comment]
That test is saying that the number of datasets created in the EXECUTION BUTLER bit of code (the first 3 pipetask calls using $NODE) is equal to the number of datasets created by the QUNTUM BACKED BUTLER bit of code (the second set of 3 pipetask calls using $NODE).

File: bin/run_demo.sh [hunk 2]
```diff
@@ -161,26 +162,26 @@ pipetask qgraph -b DATA_REPO/butler.yaml \
...
-NODES=( $(pipetask qgraph -b "DATA_REPO/butler.yaml" -g "$graph_file" --show-qgraph-header \
-    |jq -r 'first(.Nodes)[][0]') )
+  # Run the init step
+  pipetask --long-log pre-exec-init-qbb "DATA_REPO/butler.yaml" "$graph_file"
 
-# Run the three quanta one at a time to ensure that we can start from a
-# clean execution butler every time.
-node=${NODES[0]}
-pipetask --long-log run-qbb --qgraph-node-id "$node" "DATA_REPO/butler.yaml" "$graph_file"
+  # Run the three quanta one at a time to ensure that we can start from a
```
Comment (timj) [inline bin/run_demo.sh:168]: This comment should not refer to execution butler. Change it to something like:

```
# Run each pipeline step in turn.
```

File: bin/run_demo.sh [hunk 3]
```diff
@@ -83,76 +83,77 @@ pipetask run -d "exposure=903342 AND detector=10" -b DATA_REPO/butler.yaml \
...
-refresh_exe() {
-  rm -rf "$tmp_butler"
-  mkdir "$tmp_butler"
-  cp "$exedir"/* "$tmp_butler/"
+  # Run the init step
+  TMP_BUTLER="./tmp_execution_butler"
+  refresh_butler $TMP_BUTLER
+  pipetask --long-log run -b "$TMP_BUTLER" -i "$incoll" --output-run "$exerun" --init-only --register-dataset-types --qgraph "$graph_file" --extend-run
+
+  # Run the three quanta one at a time to ensure that we can start from a
```
Comment (timj) [inline bin/run_demo.sh:117]: We are not running serially "to ensure" that.  The original comment had a comma to indicate that the fresh butler was used each time. How about:

```
# Run the three quanta sequentially with a refreshed execution butler ever time.
```

File: bin/run_demo.sh [hunk 4]
```diff
@@ -83,76 +83,77 @@ pipetask run -d "exposure=903342 AND detector=10" -b DATA_REPO/butler.yaml \
...
+  # clean execution butler every time.
+  for NODE in $(pipetask qgraph -b "$TMP_BUTLER" -g "$graph_file" --show-qgraph-header \
+      |jq -r 'first(.Nodes)[][0]')
+  do
+    # Run the execution butler in multiple steps, ensuring that a fresh
+    # butler is used each time.
+    TMP_BUTLER_NODE="./tmp_execution_butler-$NODE"
+    refresh_butler $TMP_BUTLER_NODE
+
+    # Run these three pipetasks concurrently
```
Comment (timj) [inline bin/run_demo.sh:127]: Remove the comment here I think since the comment for the loop is clear what is happening.

File: bin/run_demo.sh [hunk 5]
```diff
@@ -83,76 +83,77 @@ pipetask run -d "exposure=903342 AND detector=10" -b DATA_REPO/butler.yaml \
...
+  TMP_BUTLER="./tmp_execution_butler"
+  refresh_butler $TMP_BUTLER
+  pipetask --long-log run -b "$TMP_BUTLER" -i "$incoll" --output-run "$exerun" --init-only --register-dataset-types --qgraph "$graph_file" --extend-run
+
+  # Run the three quanta one at a time to ensure that we can start from a
+  # clean execution butler every time.
+  for NODE in $(pipetask qgraph -b "$TMP_BUTLER" -g "$graph_file" --show-qgraph-header \
+      |jq -r 'first(.Nodes)[][0]')
+  do
+    # Run the execution butler in multiple steps, ensuring that a fresh
```
Comment (timj) [inline bin/run_demo.sh:122]: This is also implied from the loop description.

**Review by timj — APPROVED** [review]
This looks good. There is a small possibility that the two parallel processes hitting sqlite would fall over each other but given that the second run is using the graph and not the sqlite at all and we only have two processes, I think this will almost certainly be fine. cc/ @andy-slac

**Pull request merged**